core/reorder/rcm_kernels.hpp

```diff
@@ -0,0 +1,119 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_REORDER_RCM_KERNELS_HPP_
+#define GKO_CORE_REORDER_RCM_KERNELS_HPP_
+
+
+#include <ginkgo/core/reorder/rcm.hpp>
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+
+
+#define GKO_DECLARE_RCM_GET_PERMUTATION_KERNEL(IndexType)                \
+    void get_permutation(                                                \
+        std::shared_ptr<const DefaultExecutor> exec,                     \
+        const size_type num_vertices, const IndexType *const row_ptrs,   \
```

Comment:
Yes, it is not part of the interface - this can also be deduced from it not showing up in IDEs hovering declarations.
I like it that way though, it helps me to reason about the code and makes the compiler shout at me in case I do something stupid.

Reply:
But in case you meant that - yes, I'll remove it from the declaration.

Reply:
It's fine, you can also leave it in. Only the mixture of west-const and east-const might be confusing for some people (also, as far as I can tell, there is no easy way to write west-const constant pointer types. This was mainly a comment on what we as a library could do better in terms of const-correctness for local variables and parameters.

Reply:
Concerning const placement, I always remember it with: "const effects what's directly left of it, unless there is nothing, then it effects what's right". On top of it, I prefer left placement, leading to the syntax `const T *const`.

Reply:
I was referring to the tongue-in-cheek conflict in the C++ community regarding the placement of `const` between proponents of `const west` and `east const`, but this only as a side note. I'm fine with how it is right now.